### ISSUE TITLE
"Reload page to use this extension" due to new documentStart script

### ISSUE THREAD

**Issue by comex** [FooSoft/yomichan#1962]
**Description**
In Chrome's permissions settings, I have Yomichan set to load "On click" rather than "On specific sites" or "On all sites".  This is not because I distrust it; it's just that I want to keep my browsing experience as light as possible.  Even if Yomichan's own on/off toggle is set to off, there's going to be some small cost to just loading a bunch of JS files into every web page.  I love Yomichan but only need it occasionally, so "On click" is perfect for me.

Previously, I could click the Yomichan icon in Chrome's toolbar, and Yomichan would start working immediately for the current page.  But as of recently, clicking the icon  causes a Chrome dialog to pop up saying "Reload page to use this extension":

<img width="465" alt="image" src="https://user-images.githubusercontent.com/47517/134824242-82f0685e-2f38-4cdc-8afd-e130f0c903b9.png">

Clicking Reload does what it says, and only then does Yomichan start working.

Upon investigation, it appears that Chrome does this for extensions that have a `document_start` content script applicable to the current page ([[1]](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/extensions/extension_action_runner.cc;l=188;drc=58ec136383357afd0692538467fb86797c5fd3d8) [[2]](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/extensions/extension_action_runner.cc;l=54;drc=58ec136383357afd0692538467fb86797c5fd3d8)).  It doesn't do the same for `document_idle` scripts.  I suppose they reasoned that a script that's normally delayed until an arbitrary 'idle' point should still work if further delayed until the user clicks a button, while a script that asks to run ["before any other DOM is constructed or any other script is run"](https://developer.chrome.com/docs/extensions/mv3/content_scripts/#run_time) isn't safe to delay in the same way.

Yomichan's main injection point is `document_idle`, but in #1875, a separate `document_start` script was added, which sends a synchronous `documentStart` message to the background page.  This then calls `_updateTabAccessibility`, which currently does nothing unless the hostname is "docs.google.com".

Since the script is only used on that one domain, one possible solution is to change the manifest to only load `document-start.js` on `https://docs.google.com/*` rather than `https://*/*`, and ditto for http.  I tested this locally and verified that it brings back the ability to load Yomichan without reloading the page.  That said, this approach may make it more difficult to add more domains in the future, since I think Chrome would require user re-approval in that case (but I'm not sure how this works exactly).  Also, the "Reload page" prompt would still be shown when visiting Google Docs.  Still, it would be a big improvement for me.

Another possible solution might be some kind of lazy-loading built into the extension itself, but I know that would be a big feature to implement.

**Browser version**
Google Chrome 94.0.4606.61 (Official Build) (x86_64) on macOS Version 11.4 (Build 20F71)

**Yomichan version**
21.8.31.1

**Comment by toasted-nutbread** [comment]
Thanks for the detailed report, I remember briefly reading about this somewhere before but didn't think it would apply to Yomichan. For now, I'll update the document-start.js script to only run for docs.google.com.

> Another possible solution might be some kind of lazy-loading built into the extension itself, but I know that would be a big feature to implement.

It's not really a huge feature, and I looked into it while investigating the Google Docs accessibility issue. I remember there being two issues:
* It requires an additional permission to watch for changes to web page navigation on all tabs. Ultimately I don't think it would be a huge deal, but I've been trying to avoid adding permissions that are only necessary for a very rare edge case.
* I did notice some functionality for declaratively adding content scripts, but I think it was Firefox only and/or it won't be supported in Chrome's MV3.

I'll make a note to revisit this at some point to see if there is a better solution to only run the script if the option is enabled.
